scalyr_agent/util.py

```diff
@@ -1389,48 +1388,57 @@ class BlockingRateLimiter(object):
     each of num_agents Scalyr Agents instantiates one of these rate limiters such that the aggregate access rate over
     all agents is R.  (Therefore, the per-agent rate is R / num_agents).
 
-    Whenever a certain number of consecutive successes are seen, the rate is increased up until upper_cluster_rate.
-    Likewise, whenever a failure is reported (on token release()), the rate is lowered up until a lower_cluster_rate.
+    Whenever a certain number of consecutive successes are seen, the rate is increased up until max_cluster_rate.
+    Likewise, whenever a failure is reported (on token release()), the rate is lowered up until a min_cluster_rate.
     Cluster rate is reduced by dividing current rate by backoff_factor.  Cluster rate is increased either by
-    multiplying by increase_factor or resetting to initial rate, depending on the increase_strategy.
+    multiplying by increase_factor or resetting to initial rate, depending on the strategy.
     (Note: if cluster rate is already greater or equal to initial rate, it will be multiplied by the increase factor)
     """
 
-    INCREASE_STRATEGY_MULTIPLY = 'multiply'
-    INCREASE_STRATEGY_RESET_THEN_MULTIPLY = 'reset_then_multiply'
+    STRATEGY_MULTIPLY = 'multiply'
+    STRATEGY_RESET_THEN_MULTIPLY = 'reset_then_multiply'
+    HARD_LIMIT_MAX_CLUSTER_RATE = 20000
```

Comment:
Good thoughts. 

I think the max should not be tied to num agents. (Conceptually, it is the max load you want to put on a single api master, regardless of num agents -- also  yesterday's conversation about not counting on other masters due to uneven load distribution means we don't count number of api masters).

Yes, we will be hobbling if num_agents is large, but that is the point of it really -- to force each agent to go really slow in a large cluster setting.

I am inclined to change this hard max on the initial rate only.  That would ensure we start at a reasonable number, while allow us to rely on backoff to probe the max rate possible.

Secondly, the code I put in place to do min(actual, target) rate should also prevent us from reaching stratospheric target rates even if the user specifies a huge max rate, thus eliminating the original problem that motivated this.

Reply:
Oh, and BTW I chose 20K because it seems to be a "traditional" limit where servers start hitting issues (may not apply today)
https://groups.google.com/forum/#!topic/dropwizard-user/QtA82VYIJwQ